File: SeriesGuide/src/main/java/com/battlelancer/seriesguide/jobs/movies/MovieWatchedJob.java
package com.battlelancer.seriesguide.jobs.movies;

import android.content.Context;
import android.support.annotation.NonNull;
import com.battlelancer.seriesguide.R;
import com.battlelancer.seriesguide.jobs.episodes.JobAction;
import com.battlelancer.seriesguide.util.MovieTools;

public class MovieWatchedJob extends MovieJob {

    private final boolean isWatched;

    public MovieWatchedJob(int movieTmdbId, boolean isWatched) {
        super(isWatched
                ? JobAction.MOVIE_WATCHED_SET : JobAction.MOVIE_WATCHED_REMOVE, movieTmdbId);
        this.isWatched = isWatched;
    }

    @Override
    protected boolean applyDatabaseUpdate(Context context, int movieTmdbId) {
        return MovieTools.setWatchedFlag(context, movieTmdbId, isWatched);
    }

    @Override
    public boolean supportsHexagon() {
        return false; // hexagon does only have watchlist and collection flags
    }

    @NonNull
    @Override
    public String getConfirmationText(Context context) {
        return context.getString(isWatched
                ? R.string.action_watched
                : R.string.action_unwatched);
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/jobs/movies/MovieWatchedJob.kt
// SPDX-License-Identifier: GPL-3.0-or-later
// SPDX-FileCopyrightText: Copyright © 2019 Uwe Trottmann <uwe@uwetrottmann.com>

package com.battlelancer.seriesguide.jobs.movies

import android.content.Context
import com.battlelancer.seriesguide.SgApp
import com.battlelancer.seriesguide.jobs.episodes.JobAction
import com.battlelancer.seriesguide.movies.tools.MovieTools

class MovieWatchedJob(
    movieTmdbId: Int,
    private val isWatched: Boolean,
    currentPlays: Int
) : MovieJob(
    if (isWatched) JobAction.MOVIE_WATCHED_SET else JobAction.MOVIE_WATCHED_REMOVE,
    movieTmdbId,
    if (isWatched) currentPlays + 1 else 0
) {

    override suspend fun applyDatabaseUpdate(context: Context, movieTmdbId: Int): Boolean {
        val movieTools = SgApp.getServicesComponent(context).movieTools()
        return if (isWatched) {
            movieTools.addToList(movieTmdbId, MovieTools.Lists.WATCHED)
        } else {
            movieTools.removeFromList(movieTmdbId, MovieTools.Lists.WATCHED)
        }
    }
}
